src/main/scala/Ch3.scala

```diff
@@ -185,5 +185,186 @@ object Ch3Ex2 {
+
+  // backward (<=) works
+  // def p2Backward[A, B, C](abOrAc: Either[A => B, A => C]): A => Either[B, C] = abOrAc match {
+  //   case Left(l: (A => B)) => { a => Left(l(a))  }
+  //   case Right(r: (A => C)) => { a => Right(r(a))  }
+  // }
+
+  // Problem 3
+  // (A * Int) + ((A * Char) + (A * Float))
+  // = (A * Int) + (A * Char) + (A * Float)
+  // = A * (Int + Char + Float)
+  type P3T1[A] = Either[(A, Int), Either[(A, Char), (A, Float)]]
+  type P3T2[A] = (A, IntOrCharOrFloat)
+
+  sealed trait IntOrCharOrFloat
+  final case class ItsInt(i: Int) extends IntOrCharOrFloat
+  final case class ItsChar(c: Char) extends IntOrCharOrFloat
+  final case class ItsFloat(f: Float) extends IntOrCharOrFloat
+
+  def p3Forward[A]: P3T1[A] => P3T2[A] = {
+    case Left((a, i)) => (a, ItsInt(i))
+    case Right(aCharFloat) => aCharFloat match {
+      case Left((a, c)) => (a, ItsChar(c))
+      case Right((a, f)) => (a, ItsFloat(f))
+    }
+  }
+
+  def p3Backward[A]: P3T2[A] => P3T1[A] = {
+    case (a, ItsInt(i)) => Left((a, i))
+    case (a, ItsChar(c)) => Right(Left((a, c)))
+    case (a, ItsFloat(f)) => Right(Right((a, f)))
+  }
+
+  // Problem 4
+  sealed trait OptEither[A, B]
+  final case class OptNone[A, B]() extends OptEither[A, B]
+  final case class OptLeft[A, B](a: A) extends OptEither[A, B]
+  final case class OptRight[A, B](b: B) extends OptEither[A, B]
+
+  def map[A, B, T](oe: OptEither[A, B])(f: B => T): OptEither[A, T] = oe match {
+    case OptRight(b) => OptRight(f(b))
+    case OptLeft(a) => OptLeft(a)
+    case OptNone() => OptNone()
+  }
+
+  def flatMap[A, B, T](oe: OptEither[A, B])(f: B => OptEither[A, T]): OptEither[A, T] = oe match {
+    case OptRight(b) => f(b)
+    case OptLeft(a) => OptLeft(a)
+    case OptNone() => OptNone()
+  }
+
+  // Problem 5
+  type MyT[T] = Boolean ⇒ MyTVals[T]
+  sealed trait MyTVals[T]
+  final case class EmptyValueT[T]() extends MyTVals[T]
+  final case class SingleValueT[T](t: T) extends MyTVals[T]
+  final case class IntWithT[T](i: Int, t: T) extends MyTVals[T]
+  final case class StringToT[T](st: String ⇒ T) extends MyTVals[T]
 
+  def mapMyT[T, U](myT: MyT[T])(f: T => U): MyT[U] = b => {
+    myT(b) match {
+      case EmptyValueT() => EmptyValueT()
+      case SingleValueT(t) => SingleValueT(f(t))
+      case IntWithT(i, t) => IntWithT(i, f(t))
+      case StringToT(sToT) => StringToT(s => f(sToT(s)))
+    }
+  }
+
+  // Map for MyTU preferring T over U
+  def mapMyTU[T, U, V](myTU: MyTU[T, U])(f: T => V): MyTU[V, U] = myTU match {
+    case EmptyValuesTU() => EmptyValuesTU()
+    case TAndU(t, u) => TAndU(f(t), u)
+    case IntAndT(i, t) => IntAndT(i, f(t))
+    case StringAndU(s, u) => StringAndU(s, u)
+  }
+
+  // Problem 6.1
+  type P6State[S, A] = S ⇒ (A, S)
+  def p6P1Map[S, A, B](sa: P6State[S, A])(f: ((S, A)) ⇒ (S, B)): P6State[S, B] = { s =>
+    val (a, s2) = sa(s)
+    val (s3, b) = f((s2, a))
+    (b, s3)
+  }
+
+  // Problem 6.2
+  // A + Z ⇒  (A ⇒  B) ⇒  B + Z
```

Comment:
`A+Z` is the disjunction, not the tuple `(A, Z)`.

Reply:
fixed!